contracts/CompoundingIndex.sol

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.2;
+
+import "./WadRayMath.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "hardhat/console.sol";
+
+/**
+  * CompoundingIndex
+  * The contract contains logic for time-based index recalculation with minimal memory footprint.
+  * It could be used as a base building block for any index-based entities like deposits and loans.
+  * @dev updatedRate the value of updated rate
+**/
+contract CompoundingIndex is Ownable {
+    using WadRayMath for uint256;
+
+    uint256 private constant SECONDS_IN_YEAR = 31536000;
+    uint256 private constant BASE_RATE = 1 ether;
+
+    uint256 public start = block.timestamp;
+
+    uint256 public index = BASE_RATE;
+    uint256 public indexUpdateTime = start;
+
+    mapping(uint256 => uint256) prevIndex;
+    mapping(address => uint256) userUpdateTime;
+
+    uint256 public rate;
+
+
+    /* ========== SETTERS ========== */
+
+
+    /**
+     * Sets the new rate
+     * Before the new rate is set, the index is updated accumulating interests
+     * @dev updatedRate the value of updated rate
+    **/
+    function setRate(uint256 _rate) public onlyOwner {
+        updateIndex();
+        rate = _rate;
+        emit RateUpdated(rate);
+    }
+
+
+    /* ========== MUTATIVE FUNCTIONS ========== */
+
+
+    /**
+     * Updates user index
+     * It persists the update time and the update index time->index mapping
+     * @dev user address of the index owner
+    **/
+    function updateUser(address user) public onlyOwner {
+        userUpdateTime[user] = block.timestamp;
+        prevIndex[block.timestamp] = getIndex();
+    }
+
+
+    /* ========== VIEW FUNCTIONS ========== */
+
+
+    /**
+     * Gets current value of the compounding index
+     * It recalculates the value on-demand without updating the storage
+    **/
+    function getIndex() public view returns(uint256) {
+      uint256 period = block.timestamp - indexUpdateTime;
+      if (period > 0) {
+        return index.wadToRay().rayMul(getCompoundedFactor(period)).rayToWad();
+      } else {
+        return index;
+      }
+    }
+
+
+   /**
+     * Gets the user value recalculated to the current index
+     * It recalculates the value on-demand without updating the storage
+    **/
+    function getIndexedValue(uint256 value, address user) public view returns(uint256) {
+        uint256 prevUserIndex = userUpdateTime[user] == 0 ? BASE_RATE : prevIndex[getLastUserUpdateTime(user)];
+
+        return value.wadToRay()
+        .rayMul(getIndex().wadToRay())
+        .rayDiv(prevUserIndex.wadToRay())
+        .rayToWad();
```

Comment:
RayToWad seems to round values, we should double check if this won't cause any issues: 
test case:
      await sut.setRate(toWei("0.5"));

      await time.increase(time.duration.years(1));
      
      let index = await sut.getIndex();
      console.log(index.toString());

      let val = await sut.getIndexedValue(1, owner.address);
      console.log(val.toString());

We've got 2, while the real answer is 1.6..., it could be safer to round it down.

Reply:
That is true, but I think this would be problematic only for really small values (input `value`), close to 1 Wei (as in example above). For values of an order of 1 ETH/AVAX (which are way closer to values used in real world scenarios) this issue doesn't arise, since resulting indexed values have similar order.
When it comes to time, even for short periods (like a minute) indexed values have still the same order as input value so this still isn't an issue.

Reply:
Test showing this scenarios:
```
    it("should set initial index 1", async function () {
      sut = await init("0.5", owner);

      await time.increase(time.duration.minutes(1));

      console.log('index: after 1 minute ', (await sut.getIndex()).toString());
      console.log('indexed 1 Wei after 1 minute: ', (await sut.getIndexedValue(1, owner.address)).toString());
      console.log('indexed 1 Ether after 1 minute: ', (await sut.getIndexedValue(toWei("1"), owner.address)).toString());

      await time.increase(time.duration.years(1));

      console.log('index after 1 year: ', (await sut.getIndex()).toString());
      console.log('indexed 1 Wei after 1 year: ', (await sut.getIndexedValue(1, owner.address)).toString());
      console.log('indexed 1 Ether after 1 year: ', (await sut.getIndexedValue(toWei("1"), owner.address)).toString());
    });
    ```
    
    Results:
    ```
index: after 1 minute  1000000951294204452
indexed 1 Wei after 1 minute:  1
indexed 1 Ether after 1 minute:  1000000951294204452
index after 1 year:  1648722832584035519
indexed 1 Wei after 1 year:  2
indexed 1 Ether after 1 year:  1648722832584035519
```

Reply:
@jakub-wojciechowski can you confirm that it's ok?

Reply:
It's ok. Please just put a comment above the function that we're aware of it @piotruda